feature/preprocessing selectors

Korijn opened the pull request:
This feature allows you to filter dependencies based on preprocessing selectors, similar to [how they work in conda-build](http://conda.pydata.org/docs/building/meta-yaml.html#preprocessing-selectors). You can supply custom selectors to the `create` and `update` commands, like so:

```yaml
name: selectors
selectors:
  - test
  - docs
dependencies:
  - python
  - pytest  # [test]
  - sphinx  # [docs]
  - doctest  # [test and docs]
```

* `conda env create` will install only `python`
* `conda env create --select test` will install `python` and `pytest`
* `conda env create --select test --select docs` will install everything
* `conda env create --select all` will install everything

Unit test included.

This also adds platform selectors, e.g.:

```yaml
name: platforms
dependencies:
  - python
  - pywin32  # [win]
```

The available platform selectors are:

* `linux`
* `linux32`
* `linux64`
* `osx`
* `win`
* `win32`
* `win64`
* `unix` (is equivalent to `osx or linux`)
* `unix32`
* `unix64`
* `x64`
* `x86`

You can specify arbitrary python expressions, since they are `eval`d, and you can use the `os` and `os.environ` (aliased as `environ`) objects as well, e.g.:

```yaml
name: platforms
dependencies:
  - python
  - pywin32  # [win32 and win64]
  - sphinx_rtd_theme  # [not (environ.get('READTHEDOCS', None) == 'True')]
```

Please note that most of the design choices were made so that the functionality mirrors that of [conda-build's preprocessing selectors](http://conda.pydata.org/docs/building/meta-yaml.html#preprocessing-selectors), so that users don't need to become familiar with different syntax.

nicoddemus commented:
Hmmm can I use selectors system based selectors like `win` or `linux`?

nicoddemus commented:
I see that you removed them in ba2e106... can you elaborate why they complicate `--select=all`?

Platform specific selectors would be a **huge** win for us, as we would not need to use Jinja preprocessing at all:

```yml
name: esss_qt10

includes:
  - {{ root }}/../plugins10core/environment.yml

dependencies:
  {% if os.environ.get('ESSS_DEBUG', '0') not in ['0', 'cpp'] %}
  - debug
  {% endif %}
  - unicodecsv==0.9.4
  {% if os.sys.platform != 'win32' %}
  - pillow==2.6.1
  {% else %}
  - pillow==2.1.0 # slightly different version copied from dist
  {% endif %}

  {% if os.environ.get('ESSS_QT_VER', '4') == '4' %}
  - pyqt==4.11.1
  - pytest-qt==1.3.0
  {% else %}
  - pyqt3==3.18.1
  {% endif %}
```

With platform-specific selectors we could us instead:

```yml
name: esss_qt10

includes:
  - {{ root }}/../plugins10core/environment.yml

dependencies:
  - debug  # [debug]
  - unicodecsv==0.9.4  
  - pillow==2.6.1  # [linux]  
  - pillow==2.1.0  # [win]
  
  - pyqt==4.11.1      # [qt4]
  - pytest-qt==1.3.0  # [qt4 and test]  
  - pyqt3==3.18.1     # [qt3]
```

Which be used with `--select test` and/or `--select debug`. The latter one looks **much** cleaner.

nicoddemus commented:
cc @tadeu

Korijn commented:
In the case of select all you have to parse the expressions and set the unknown variables to True. You can't just return everything as is brcause that would ignore the platform selectors...

Korijn commented:
Maybe I could catch the NameError, set the missing variable to True and keep restarting the algorithm untill no more NameErrors occur when --select all is specified...

nicoddemus commented:
What if we explicitly add selectors (other than platform specific ones) to the `environment.yml` file? 

```yml
name: selectors
selectors:
  - test
  - docs
dependencies:
  - python
  - pytest  # [test]
  - sphinx  # [docs]
  - doctest  # [test and docs]
```

This makes it explicit which selectors are available for use, and also allows us to implement a `list-selectors` command which lists the selectors from a `environment.yml` file.

```
$ conda env list-selectors
test
docs
```

Which is useful so other tools can query which selectors are available without resorting to parse the yaml file.

That also simplifies the implementation in the sense that it can do a stricter error checking when user passes an unknown selector in the command line.

Korijn commented:
Sounds like a plan to me!

nicoddemus commented:
Nice! :+1: 

Just to be clear, you plan to update the PR with the explicit selectors?

Korijn commented:
Yes, exactly.

tadeu commented:
This looks very good! I think that line selectors are better than `jinja2`, the file looks cleaner, and it is more simple to parse externally (less dependencies, no code execution).

For example, I'd like to make a tool to list the versions only for `compile` dependencies, so that a recompilation is triggered only when a `compile` dependency is changed.

Korijn commented:
Would that be an external tool or also part of a conda-env command?

nicoddemus commented:
@Korijn, @tadeu means an external tool. :smile:

tadeu commented:
At first I thought about an external tool, but it could be a parameter to `conda env export`, such as `conda env create --select compile`.

The only problem is that there is an intersection between `conda env export` and `conda list` commands, and the latter is outside `conda-env`.

nicoddemus commented:
> For example, I'd like to make a tool to list the versions only for compile dependencies, so that a recompilation is triggered only when a compile dependency is changed.

@tadeu great, I had a similar idea.

I think we can add also add a `conda env preprocess` command which outputs the full processed file with selectors stripped:

```
$ conda env preprocess --select test
name: selectors
selectors:
  - test
  - docs
dependencies:
  - python
  - pytest

$ conda env preprocess --select docs
name: selectors
selectors:
  - test
  - docs
dependencies:
  - python
  - sphinx
```

this way external tools can make use of the complete dependency information without duplicating the selector logic, and also taking in account platform specific selectors. This idea is out of scope as far as this PR is concerned though.

tadeu commented:
@nicoddemus I like that (I agree, we are getting "out-of-topic", but these are good ideas that are related to the one in this PR).

It would be good to have a common preprocessing code, because it may be further improved. For example, sometimes lists may become empty if all of its items are "unselected", and the YAML may become malformed. The preprocessor could remove the entire list in this case.

Hmm, isn't `select_lines` already implemented in `conda` by the way?

nicoddemus commented:
> Hmm, isn't select_lines already implemented in conda by the way?

I think it is implemented in `conda-build`.

Korijn commented:
It is, but I am not sure if importing it is such a good idea.. what if conda's interface changes?

Korijn commented:
Also, I think it depends on another function that in turn has incompatible dependencies on conda build.

Korijn commented:
Updated this to accept custom selectors. Please look at the enclosed test case. Also added platform selectors back in. Apologies for the delay!

@nicoddemus @msarahan Please review?

nicoddemus commented on tests/test_env.py:
```diff
@@ -43,6 +43,42 @@ def test_with_pip(self):
         self.assert_('foo' in e.dependencies['pip'])
         self.assert_('baz' in e.dependencies['pip'])
 
+    def test_with_selectors(self):
```
Could you also check one of the platform selectors in this test? I think adding:

```
  - pywin32 # [win]
```

And adding an extra test like this:

```python
if sys.platform.startswith('win'):
    self.assert_('pywin32' in e.dependencies['conda'])
else:
    self.assert_('pywin32' not in e.dependencies['conda'])
```

would be enough

Korijn replied:
Sure, I'll also add one for architecture.

Korijn replied:
Done!

nicoddemus commented on conda_env/env.py:
```diff
@@ -58,11 +60,81 @@ def from_yaml(yamlstr, **kwargs):
+    return d
+
+
+sel_pat = re.compile(r'(.+?)\s*(#.*)?\[(.+)\](?(2).*)$')
```
I have some questions about the regular expression:

* Why is the `#` optional? regex:  `(#.*)?`
* Why matching again against group `2` at the end of the regex?  regex: `(?(2).*)`

Korijn replied:
I'll be honest here: I ripped it from conda-build, assuming that it's probably safe since it's been exposed to a lot of usage already. I am an experienced regex user, but this is a mystery for me as well. Perhaps there is some use case for this that I'm not aware of...

Korijn replied:
I looked up the changeset that introduced the regex, and was surprised to learn that these selectors originally weren't embedded in comments... so I guess it's a bit of legacy?

https://github.com/conda/conda-build/commit/ccb5d729683855a037500b11c654a9abb467fbc1

Korijn replied:
Guess what, you can/could use selectors without comments, probably before they switched to pyyaml? https://github.com/conda/conda-recipes/blob/master/matplotlib/meta.yaml

Korijn replied:
I'll adjust the regex since we don't need to support such legacy cases.

nicoddemus replied:
Yeah, @tadeu mentioned it to me in person. I agree with you that it seems to be supported only for legacy reasons.

Since this is a new feature, I would rather to be a little more strict and always require `#` here, but that's just my opinion.

groutr replied:
Actually, I'd be in favor of the opposite (always require selector outside of comment).  We shouldn't expect comments to survive for very long in a .yaml file (especially any file read with pyyaml).  Requiring selectors to be in comments makes the use case of automatically modifying recipes via software almost impossible to do (since the very first bit of information thrown away are the comments).  The selectors were originally (and still are, I believe) expected to be the last thing on a line (whether in a comment or not).

msarahan replied:
But then we get back to whether these files are actually parsable yaml, right? What makes the files most standard compatible and least lossy?

Korijn replied:
Personally, I think the approach should be compatible with what conda-build users are already familiar with. Inconsistent syntax/API is incredibly confusing.

In fact, I fell for that trap this week on appveyor, where you can expand environment variables with windows %syntax% everywhere, except for the artifacts $(section).

I also disagree with "almost impossible". The code in this PR accomplishes it, and it's not hard at all.

On another note, the `(?(2).*)$` part of the regex is a conditional part: if group 2 (`(#.*)?`) matched, additional characters are allowed before the end of the string. In other words, *if* the selector is not in a comment, additional characters are *not* allowed after the selector.

Korijn replied:
I think there's no harm in continuing to support both formats, to keep both parties in this discussion happy. So, to summarize, this is now considered valid:

```
name: myenv
dependencies:
  - python
  - pywin32  # [win] we need this on windows
  - py2exe  [win]
```

And this is invalid:

```
name: myenv
dependencies:
  - python
  - pywin32  [win] we need this on windows
```

nicoddemus commented:
@Korijn thanks a lot for continuing to pursue this! :smile:

Korijn commented:
I think we're good to go here.

nicoddemus commented:
:+1:

Korijn commented:
@msarahan can this be reviewed and possibly be merged in? I've put a lot of time and effort into this PR so it'll satisfy a large number of users. Hoping to get it in soon.

tadeu commented:
Ping @kalefranz

msarahan commented:
LGTM. Nice improvements.  Leaving it to @kalefranz for final say.

Korijn commented:
One more ping @msarahan and @kalefranz for this, I'd really like to get this in! :)

msarahan commented:
Thanks @Korijn, sorry for the delay.  We'll try to get this into a release soon.
